### ISSUE TITLE
NPM Build Time is Too Slow

### ISSUE THREAD

**Issue by ShayNehmad** [guardicore/monkey#608]
## Describe the bug
Building the UI (which means running `` and `npm run dist` takes way too long. 

## To Reproduce
1. cd into the cc/UI dir
2. Build the UI
3. Go make a cup of coffee 🙄

## Expected behavior
Build time should be measured in seconds, not minutes. 

## Tasks
- [ ] Improve build time (Time boxed 0.25d) @ilija-lazoroski 
    - [x] Create new issues with ideas about how to further improve the performance

**Comment by VakarisZ** [comment]
Can't we just ship island with a built UI?

**Comment by ilija-lazoroski** [comment]
Measurement on what we need to focus on:
<img width="336" alt="image" src="https://user-images.githubusercontent.com/15820737/162732976-d8a57f8d-0175-45f0-8876-039f38a8ead0.png">

**Comment by ilija-lazoroski** [comment]
One approach is to use [HappyPack](https://www.npmjs.com/package/happypack) which the latest release is done in 2019 and the person maintaining it, lost interest in the project. The improvement is  ~60% for production and ~71% for development on my machine.

Production (`npm run dist`):
![image](https://user-images.githubusercontent.com/15820737/162772068-98b6317b-2d8b-42a4-876f-0f83c7f718e4.png)

Development (`npm run dev`):
![image](https://user-images.githubusercontent.com/15820737/162773359-adc34227-e71d-4d44-9ad3-41ca0a3032f6.png)


The **webpack.config.js** now looks like below:
```
const path = require('path');

const HtmlWebPackPlugin = require("html-webpack-plugin");
const SpeedMeasurePlugin = require("speed-measure-webpack-plugin");
const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin');
const HappyPack = require('happypack');
var happyThreadPool = HappyPack.ThreadPool({ size: 10 });
const smp = new SpeedMeasurePlugin();

module.exports = smp.wrap({
  module: {
    rules: [
      { test: /\.tsx?$/,
        loader: 'ts-loader',
        options: {
          transpileOnly: true,
          happyPackMode: true
        }
      },
      {
        test: /\.js$/,
        use: 'happypack/loader?id=sml',
        enforce: "pre"
      },
      {
        test: /\.js$/,
        use: 'happypack/loader?id=jsx',
        exclude: /node_modules/,
      },
      {
        test: /\.css$/,
        use: 'happypack/loader?id=css'
      },
      {
        test: /\.scss$/,
        use: 'happypack/loader?id=scss'
      },
      {
        test: /\.(ttf|eot|svg)(\?v=[0-9]\.[0-9]\.[0-9])?$/,
        type: 'asset/resource'
      },
      {
        test: /\.woff(2)?(\?v=[0-9]\.[0-9]\.[0-9])?$/,
        type: 'asset'
      },
      {
        test: /\.(png|jpg|gif)$/,
        type: 'asset'
      },
      {
        test: /\.html$/,
        use: [
          {
            loader: "html-loader"
          }
        ]
      }
    ]
  },
  devtool: process.env.WEBPACK_MODE === 'production' ? "source-map" : "eval",
  plugins: [
    new HtmlWebPackPlugin({
      template: "./src/index.html",
      filename: "./index.html"
    }),
    new ForkTsCheckerWebpackPlugin({
      typescript: {
        diagnosticOptions: {
          semantic: true,
          syntactic: true,
        },
      },
    }),
    new HappyPack({
      id: 'jsx',
      threadPool: happyThreadPool,
      loaders: ['babel-loader']
    }),
    new HappyPack({
      id: 'sml',
      threadPool: happyThreadPool,
      loaders: ['source-map-loader']
    }),
    new HappyPack({
      id: 'css',
      threadPool: happyThreadPool,
      loaders: ['style-loader', 'css-loader']
    }),
    new HappyPack({
      id: 'scss',
      threadPool: happyThreadPool,
      loaders: ['style-loader', 'css-loader', 'sass-loader']
    }),
  ],
  resolve: {
    extensions: ['.ts', '.tsx', '.js', '.jsx', '.css'],
    modules: [
      'node_modules',
      path.resolve(__dirname, 'src/')
    ]
  },
  optimization: {
    splitChunks: {
      chunks: 'all',
    },
  },
  output: {
    publicPath: '/'
  },
  devServer: {
    host: '0.0.0.0',
    proxy: {
      '/api': {
        target: 'https://localhost:5000',
        secure: false,
        changeOrigin: true
      }
    }
  }
});
```

**Comment by ilija-lazoroski** [comment]
Using [`thread-loader`](https://www.npmjs.com/package/thread-loader) we got even better results. Because it is 'well' maintained I will go with this approach. Addition to the above is `CacheDirectory` in [`babel-loader`](https://www.npmjs.com/package/babel-loader) which will be used when babel first runs.

The time reduction is ~63% for production and ~73% for development on my machine.

Production (`npm run dist`):
![image](https://user-images.githubusercontent.com/15820737/162790355-87556387-3d3e-4b32-8bb7-2f761e498373.png)


Development (`npm run dev`):
![image](https://user-images.githubusercontent.com/15820737/162790431-bcabe8b7-1a3f-4796-b646-13982c5c1fde.png)

**Comment by ilija-lazoroski** [comment]
For further improvement we can look at [`DLLPlugin`](https://webpack.js.org/plugins/dll-plugin/)

**Comment by ilija-lazoroski** [comment]
Times after caching on production: 
![image](https://user-images.githubusercontent.com/15820737/162799062-537e8f99-6582-4ccf-bfe2-8a958bb2d5b6.png)

**Comment by ilija-lazoroski** [comment]
AppImage build time using `npm run dist` using `devtool: source-map`:
![image](https://user-images.githubusercontent.com/15820737/162975124-ff039847-da12-4f84-a719-b9a29515fc24.png)

AppImage build time using `npm run dev` using `devtool: eval-source-map`:
![image](https://user-images.githubusercontent.com/15820737/162975252-ea4f4dab-055a-47b3-8da4-ae9f64f15af0.png)


Docker build time using `npm run dist` using `devtool: source-map`:
![image](https://user-images.githubusercontent.com/15820737/162975583-e53cca3e-e956-42c5-b9d6-07dfd43ebc33.png)

Docker build time using `npm run dev` using `devtool: eval-source-map`:
![image](https://user-images.githubusercontent.com/15820737/162975463-dff11203-f544-4706-9bbd-ff83bb11dc40.png)


MSI build time using `npm run dist` using `devtool: source-map`:
![image](https://user-images.githubusercontent.com/15820737/162977241-efc82463-5a91-4a90-9325-4f873bd22c0a.png)


MSI build time using `npm run dev` using `devtool: eval-source-map`:
![image](https://user-images.githubusercontent.com/15820737/162977153-2fe13a37-062b-45df-b3ac-c4a503c7b4ed.png)

**Comment by ilija-lazoroski** [comment]
Closed by #1880 and created issues: #1878 and #1879.
